containers/src/Data/Graph.hs

```diff
@@ -166,47 +195,72 @@ deriving instance Generic1 SCC
 -- | @since 0.5.9
 deriving instance Generic (SCC vertex)
 
+-- There is no instance Lift (NonEmpty v) before template-haskell-2.15.
+#if MIN_VERSION_template_haskell(2,15,0)
 -- | @since 0.6.6
 deriving instance Lift vertex => Lift (SCC vertex)
+#else
+instance Lift vertex => Lift (SCC vertex) where
+  lift (AcyclicSCC v) = AppE <$> [| AcyclicSCC |] <*> lift v
+  lift (CyclicSCC' vs) = AppE <$> [| CyclicSCC' |] <*> liftNonEmpty vs
+
+liftNonEmpty :: Lift vertex => NonEmpty vertex -> Q Exp
+liftNonEmpty (x :| xs) = do
+  let nonemptyName = mkNameG DataName "base" "GHC.Base" ":|"
+  x' <- lift x
+  xs' <- lift xs
+  return (InfixE (Just x') (ConE nonemptyName) (Just xs'))
+#endif
```

Comment:
This looks way too complicated. `TemplateHaskellQuotes` has been around since 8.0; shouldn't you be able to use that here instead of all this low-level fiddling?